### PR TITLE
refactor(core): alpha.7 rewrite

### LINKED ISSUE #438: Support arbitrary Notary signature schemes
Right now we only support ECDSA signatures from the Notary. We need to add a layer of abstraction to support _any_ kind of signature scheme using dynamic dispatch on some new traits.

### PR DESCRIPTION
**Pull request by sinui0**
This PR completely rewrites `tlsn-core` for our alpha.7 release. Closes #438 

There are still a couple of loose ends that I'm tidying up but the vast majority of the deliverable is there. Prior to accepting this PR we will need to review our new fork of [`rs_merkle`](https://github.com/tlsnotary/rs-merkle) which was required to support hash algorithm customization.

A follow up PR will be provided which completes the integration of these changes into the other crates. I have most of this integration work complete on a separate branch already.

# Highlights

- Improved the structure of attestations
  - Header has a static smol size and now includes a version tag and unique identifier
  - Body fields are now merkelized so that only fields necessary to prove desired statements need to be included
  - Body is forward-compatible with new kinds of fields (for example, new kinds of commitments)
- Added a `Request` pattern which will allow a Prover to configure aspects of the attestation depending on what the Notary supports.
  - This paves the way for custom extensions, eg #342 
- Further encapsulated various kinds of proofs
  - `ServerIdentityProof` proves the server name using the certificate chain
  - `AttestationProof` proves the Notary signature and attestation body
  - `TranscriptProof` proves subsequences of the transcript using the encoding or hash commitments.
- Extensible cryptographic algorithms via `CryptoProvider` (inspired by `rustls`).
  - Supports multiple hash algorithms, including user provided custom.
  - Supports multiple signature algorithms, including user provided custom.
- Added a new `PlaintextHash` commitment type (not fully exposed yet) which will be implemented with authdecode. See #479 
- Encapsulated more logic into this crate in the form of builders, reducing API surface area and making it more difficult to misuse. There should now be very little implementation code in `prover` and `verifier` crates related to attestations.

# Not Included

There are some items that I initially wanted in scope but decided to remove as they proved to be larger than anticipated.

- Protobufs
- Canonical serialization format which is efficient to implement in the EVM. For now we continue to use BCS which uses varint encoding. This may actually be ok, but I've not looked into it.
- Extensive versioning
- #342 
- #351

### CODE REVIEW

File: crates/core/tests/api.rs [hunk 1]
```diff
@@ -1,192 +1,110 @@
...
-        testdata.dns_name.as_str()
-    );
+    let transcript_proof = transcript_proof_builder.build().unwrap();
 
-    let (sent, recv) = substrings_proof.verify(&header).unwrap();
+    // Verifier verifies the proofs.
 
-    assert_eq!(&sent.data()[range1], b"se".as_slice());
-    assert_eq!(&recv.data()[range2], b"ec".as_slice());
+    todo!()
```
Comment (th4s) [inline crates/core/tests/api.rs:109]: todo

**Review by th4s — COMMENTED** [review]
:rocket: Great work :rocket:

Overall looking very good! Most of the complexity is encapsulated and organized well into  modules. Apart from the remaining `todo`s and clippy comments I think the only thing needed is more unit tests.

File: crates/core/src/signing.rs [hunk 2]
```diff
@@ -0,0 +1,261 @@
...
+
+    use k256::ecdsa::{
+        signature::{SignerMut, Verifier},
+        Signature as Secp256K1Signature, SigningKey,
+    };
+
+    use super::*;
+
+    /// secp256k1 signer.
+    pub struct Secp256k1Signer(Arc<Mutex<SigningKey>>);
```
[the remark this answers is not in the document]
Comment (th4s) [reply crates/core/src/signing.rs:215]: Hmm that would not have come to my mind in this situation. Do you expect the `SigningKey` to change? Or is it about not having sensitive information distritbuted in RAM?

**Comment by themighty1** [comment]
I have e general question while still re-viewing this PR. Some time ago we started using the "TLS peer" terminology since we allow for the TLS Server to act as a TLSN Prover.
This PR goes back to using the "Prover==TLS Client" terminology, e.g. by calling the type "ServerCertCommitment" instead of "TLSPeerCertCommitment". 
Is that intended?

**Comment by maceip** [comment]
>   * `ServerIdentityProof` proves the server name using the certificate chain
>   * `AttestationProof` proves the Notary signature and attestation body
>   * `TranscriptProof` proves subsequences of the transcript using the encoding or hash commitments.

👨‍🍳

File: crates/core/src/presentation.rs [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+    CryptoProvider,
+};
+
+/// A verifiable presentation.
+#[derive(Debug)]
+pub struct Presentation {
+    attestation: AttestationProof,
+    identity: Option<ServerIdentityProof>,
+    transcript: Option<TranscriptProof>,
+}
```
Comment (maceip) [inline crates/core/src/presentation.rs:18]: this feels like the right place for a 'quote' from the hardware, when available. todo for me  @maceip 

```diff
+   measurement: Option<HardwareReport>
```

File: crates/core/src/signing.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (th4s) [reply crates/core/src/signing.rs:219]: yeah if the api requires it, no problem.

**Review by th4s — APPROVED** [review]
:rocket: nice

**Review by themighty1 — COMMENTED** [review]
.

**Comment by themighty1** [comment]
ACK, including out rs-merkle fork, (will open a PR with some fixes)
@sinui0 could you pls answer my q above.

**Pull request merged**